File: ngo-backend/src/models/CaptureTiming.js
const mongoose = require('mongoose');

/**
 * Diagnostic-only: the four timestamps that describe how long a captured
 * payment took to reach us, and where the time went.
 *
 * Written by POST /api/apk/capture-timing (routes/captureTiming.js), which the
 * phone calls alongside the real capture upload. Deliberately a separate
 * collection and a separate request rather than extra fields on RawEvent: this
 * is a temporary investigation into the reported 10-14 minute gap, and it
 * should be removable without touching the capture path that moves money.
 *
 * Joined back to the real RawEvent on (deviceId, utcTimestamp) — utcTimestamp
 * is the notification's own post time, formatted identically by the same
 * TimeFormatter on both uploads, so it identifies the same capture.
 */
const captureTimingSchema = new mongoose.Schema(
  {
    deviceId: { type: String, required: true, index: true },
    traderId: { type: Number, default: null, index: true },
    /** "NOTIFICATION" or "SMS" — the two capture paths under comparison. */
    captureSource: { type: String, default: '' },
    /** Android package for notifications; the SMS sender address for SMS. */
    origin: { type: String, default: '' },

    // ---- the four timestamps ----
    /**
     * 1. Embedded — the time written INSIDE the message text ("at 11:05 pm"),
     * which is what the trader reads and compares against. Stored raw as well
     * as parsed: the raw string is evidence, the epoch is only our reading of
     * it, and a parse that guesses wrong must not be able to masquerade as
     * data.
     */
    embeddedTimeText: { type: String, default: '' },
    embeddedTimeMs: { type: Number, default: null },
    /**
     * 2. System post — Android's own record of when the notification was
     * posted (StatusBarNotification.getPostTime), or for SMS the telephony
     * layer's received timestamp (SmsMessage.getTimestampMillis). Independent
     * of our code entirely.
     */
    systemPostTimeMs: { type: Number, default: null },
    /**
     * 3. App reaction — wall clock at the instant our callback actually ran
     * (onNotificationPosted / SMSReceiver.onReceive). The gap from 2 to 3 is
     * the OS holding the event back from us.
     */
    appReactionTimeMs: { type: Number, default: null },
    /** 4. Server received — stamped here, on arrival. */
    serverReceivedAt: { type: Date, default: Date.now },

    /** The join key back to RawEvent, and enough text to identify the event. */
    utcTimestamp: { type: String, default: '', index: true },
    bodyPreview: { type: String, default: '' },
    amount: { type: String, default: '' },

    // ---- duplicate-capture investigation (notifications only) ----
    /**
     * StatusBarNotification.getKey() — Android's stable identity for a
     * notification. Two capture-timing rows with the SAME notifKey are the SAME
     * underlying notification delivered to our listener twice (e.g. redelivered
     * after a listener rebind — see HeartbeatService.requestRebind), NOT a
     * second post by GPay. Empty for SMS.
     */
    notifKey: { type: String, default: '', index: true },
    notifId: { type: Number, default: null },
    notifTag: { type: String, default: '' },
    /** getGroupKey() — distinguishes a group summary from its children. */
    groupKey: { type: String, default: '' },
    /**
     * getPostTime() as read on the callback, BEFORE the onNotificationPosted
     * `<=0 -> System.currentTimeMillis()` fallback. When this is 0, the
     * systemPostTimeMs above is our own wall clock, not GPay's post time — which
     * is exactly how a redelivered notification (unchanged original post time,
     * or a reconstructed one reporting 0) can look ~minutes newer than the first.
     */
    rawPostTimeMs: { type: Number, default: null },

    createdAt: { type: Date, default: Date.now },
  },
  { timestamps: true }
);

module.exports = mongoose.model('CaptureTiming', captureTimingSchema);

File: apk/app/src/main/java/com/example/paymentbot/CaptureTiming.java
package com.example.paymentbot;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import org.json.JSONObject;

import java.util.Calendar;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Diagnostic instrumentation for the reported 10-14 minute capture delay.
 *
 * Records the three timestamps only the phone can know, for every captured
 * notification and SMS, and ships them to POST /api/apk/capture-timing:
 *
 *   1. Embedded      — the time written inside the message text ("at 11:05 pm").
 *                      What the trader reads. Not a system clock at all: the
 *                      sending app chose it.
 *   2. System post   — StatusBarNotification.getPostTime() for notifications,
 *                      SmsMessage.getTimestampMillis() for SMS. Android's own
 *                      record, independent of this app.
 *   3. App reaction  — wall clock at the instant our callback ran.
 *
 * The fourth (server received) is stamped server-side on arrival.
 *
 * Deliberately separate from the capture upload rather than extra fields on
 * it: this is temporary, and it must not be able to change, delay or fail the
 * event that actually settles a payment. Every method here swallows its own
 * errors for the same reason.
 */
public final class CaptureTiming {

    private static final String TAG = "PaymentBot";
    private static final String ENDPOINT = "/api/apk/capture-timing";

    /**
     * Times as the payment apps write them: "at 11:05 pm", "at 6:47 PM",
     * "on 12-08-26 at 21:48", "11:05pm". Hour and minute are what matter —
     * the reported gap is minutes, so seconds are noise.
     */
    private static final Pattern EMBEDDED_TIME = Pattern.compile(
            "\\b(?:at\\s+)?([0-9]{1,2}):([0-9]{2})\\s*([ap])\\.?m\\.?", Pattern.CASE_INSENSITIVE);
    /** 24-hour fallback for bank SMS ("on 12-08-26-21:48"). */
    private static final Pattern EMBEDDED_TIME_24H = Pattern.compile(
            "\\b([01]?[0-9]|2[0-3]):([0-5][0-9])\\b");

    private CaptureTiming() {
    }

    /** The raw embedded-time text, exactly as written, or "" if there is none. */
    public static String embeddedText(String body) {
        if (TextUtils.isEmpty(body)) return "";
        Matcher m = EMBEDDED_TIME.matcher(body);
        if (m.find()) return m.group().trim();
        Matcher m24 = EMBEDDED_TIME_24H.matcher(body);
        if (m24.find()) return m24.group().trim();
        return "";
    }

    /**
     * The embedded time resolved to epoch millis, using {@code referenceMs}
     * (the system post time) for the calendar date the text omits.
     *
     * Returns 0 when there is nothing parseable — never a guess. If the
     * resolved time lands more than an hour AHEAD of the reference, it is
     * read as the previous day rather than the future: a notification cannot
     * describe a payment that has not happened yet, and around midnight the
     * date carried over is the only thing that can be wrong.
     */
    public static long embeddedMillis(String body, long referenceMs) {
        if (TextUtils.isEmpty(body) || referenceMs <= 0) return 0L;

        int hour = -1;
        int minute = -1;

        Matcher m = EMBEDDED_TIME.matcher(body);
        if (m.find()) {
            try {
                hour = Integer.parseInt(m.group(1)) % 12;
                minute = Integer.parseInt(m.group(2));
                if ("p".equalsIgnoreCase(m.group(3))) hour += 12;
            } catch (Exception ignored) {
                return 0L;
            }
        } else {
            Matcher m24 = EMBEDDED_TIME_24H.matcher(body);
            if (m24.find()) {
                try {
                    hour = Integer.parseInt(m24.group(1));
                    minute = Integer.parseInt(m24.group(2));
                } catch (Exception ignored) {
                    return 0L;
                }
            }
        }
        if (hour < 0 || minute < 0) return 0L;

        Calendar c = Calendar.getInstance();
        c.setTimeInMillis(referenceMs);
        c.set(Calendar.HOUR_OF_DAY, hour);
        c.set(Calendar.MINUTE, minute);
        c.set(Calendar.SECOND, 0);
        c.set(Calendar.MILLISECOND, 0);
        long resolved = c.getTimeInMillis();
        if (resolved > referenceMs + 60 * 60 * 1000L) {
            resolved -= 24 * 60 * 60 * 1000L;
        }
        return resolved;
    }

    /**
     * Queues one timing record. Goes through EventQueue like every other
     * upload, so a diagnostic never opens its own connection on a capture
     * callback and a record is not lost if the device is briefly offline.
     *
     * @param utcTimestamp the same formatted post time the real capture upload
     *                     sends — the key the two are joined on server-side.
     */
    public static void record(Context context, String captureSource, String origin,
                              String body, String amount, long systemPostMs,
                              long appReactionMs, String utcTimestamp) {
        // SMS path has no Android notification identity — rawPostTime is just the
        // telephony timestamp (no <=0 fallback distinction there), and none of
        // the group-summary/extras-dump/custom-view diagnostic fields apply to
        // SMS either.
        record(context, captureSource, origin, body, amount, systemPostMs,
                appReactionMs, utcTimestamp, "", systemPostMs, 0, "", "",
                false, "", "", "", "", "", "",
                false, false, "", "");
    }

    /**
     * Full form, used by the notification path. Adds the notification's Android
     * identity and its RAW post time (before onNotificationPosted's {@code <=0}
     * fallback), so the server can tell a redelivered notification — same
     * {@code notifKey} after a listener rebind — from a genuinely new GPay post,
     * and see whether the uploaded timestamp came from GPay or from our fallback
     * wall clock. Part of the duplicate-capture investigation.
     *
     * TEMPORARY — isGroupSummary through summaryText are the Paytm Business
     * investigation; customViewAttempted through customViewError are BUG-57's
     * confirmation signal (did the RemoteViews-inflation fallback actually work,
     * and what did it read). Both route real-device data through this same
     * server upload since the team verifying them has no adb access. Remove
     * each once its question is answered for good and the fix is trusted
     * without per-event confirmation.
     */
    public static void record(Context context, String captureSource, String origin,
                              String body, String amount, long systemPostMs,
                              long appReactionMs, String utcTimestamp,
                              String notifKey, long rawPostTimeMs, int notifId,
                              String notifTag, String groupKey,
                              boolean isGroupSummary, String extrasDump,
                              String title, String text, String bigText,
                              String subText, String summaryText,
                              boolean customViewAttempted, boolean customViewSucceeded,
                              String customViewExtractedText, String customViewError) {
        try {
            JSONObject json = new JSONObject();
            json.put("captureSource", captureSource == null ? "" : captureSource);
            json.put("origin", origin == null ? "" : origin);
            json.put("embeddedTimeText", embeddedText(body));
            json.put("embeddedTimeMs", embeddedMillis(body, systemPostMs));
            json.put("systemPostTimeMs", systemPostMs);
            json.put("appReactionTimeMs", appReactionMs);
            json.put("utcTimestamp", utcTimestamp == null ? "" : utcTimestamp);
            json.put("bodyPreview", body == null ? "" : (body.length() > 300 ? body.substring(0, 300) : body));
            json.put("amount", amount == null ? "" : amount);
            // Duplicate-capture investigation — notification identity + raw post
            // time. Empty/0 for SMS captures.
            json.put("notifKey", notifKey == null ? "" : notifKey);
            json.put("notifId", notifId);
            json.put("notifTag", notifTag == null ? "" : notifTag);
            json.put("groupKey", groupKey == null ? "" : groupKey);
            json.put("rawPostTimeMs", rawPostTimeMs);
            // TEMPORARY — Paytm Business investigation.
            json.put("isGroupSummary", isGroupSummary);
            json.put("extrasDump", extrasDump == null ? "" : extrasDump);
            json.put("resolvedTitle", title == null ? "" : title);
            json.put("resolvedText", text == null ? "" : text);
            json.put("resolvedBigText", bigText == null ? "" : bigText);
            json.put("resolvedSubText", subText == null ? "" : subText);
            json.put("resolvedSummaryText", summaryText == null ? "" : summaryText);
            // TEMPORARY — BUG-57 confirmation signal.
            json.put("customViewAttempted", customViewAttempted);
            json.put("customViewSucceeded", customViewSucceeded);
            json.put("customViewExtractedText", customViewExtractedText == null ? "" : customViewExtractedText);
            json.put("customViewError", customViewError == null ? "" : customViewError);
            EventQueue.enqueue(context, ENDPOINT, json.toString(), true);

            // Also on-device, so the same numbers are readable over adb even
            // if the phone never reaches the server.
            Log.i(TAG, "TIMING " + captureSource + " embedded=" + embeddedText(body)
                    + " post=" + systemPostMs + " rawPost=" + rawPostTimeMs
                    + " app=" + appReactionMs + " postToApp=" + (appReactionMs - systemPostMs) + "ms"
                    + " origin=" + origin + " key=" + notifKey);
        } catch (Exception e) {
            Log.e(TAG, "CaptureTiming.record failed: " + e.getMessage());
        }
    }
}
